src/devices/interlogix.c

```diff
@@ -0,0 +1,392 @@
+ *      bits and device bits provide a set of data bits.
+ * Bits 40 through 42 provide a 3-bit trigger count that is incremented for each group of message
+ *      packets. The trigger count is a simple but effective way for preventing a third party from
+ *      recording a message packet transmission and then re-transmitting that message packet
+ *      transmission to make the system controller think that a valid message packet is being transmitted.
+ * Bit 43 provides the low battery bit.
+ * Bits 44 through 53 provide the latch bit value and the debounced value for each of the five inputs
+ *      associated with the transmitter. For the F5 input, both a positive and negative latch bit are provided.
+ * Bit 55 provides even parity over odd bits 15 to 55.
+ * Bit 56 provides odd parity over even bits 16 to 56.
+ * Bit 57 is a programmable bit that can be used for a variety of applications, including providing an
+ *      additional bit that could be used for the sensor identification code or device type code.
+ * Bit 58 is a 366 ms RF on signal that functions as the "old" stop bit. This bit provides compatibility with
+ *      prior system controllers that may be programmed to receive a 58-bit message.
+ * Bit 59 is a logical 1.
+ * Bits 60 to 62 are a modulus eight count of the number of 1 bits in bits 15 through 54, providing enhanced
+ *      error detection information to be used by the system controller. Finally, bit 63 is the "new" stop bit,
+ *      e.g., a logical 0, that tells the system controller that it is the end of the message packet.
+ *
+ *
+ *
+ */
+
+#define INTERLOGIX_CRC_POLY    0x07
+#define INTERLOGIX_CRC_INIT    0x00
+
+//preamble (and single sync bit) should be 0000 0000 0000 1
+//TODO: only searching for 0000 0001 and should be searching for all 13 bits of preamble
+static unsigned char preamble[1] = { 0x01 };
+
+static uint8_t reverse4(uint8_t  nib)
+{
+        int revnib = 0;
+
+        revnib += (nib >> 3) & 0x1;
+        revnib += (nib >> 1) & 0x2;
+        revnib += (nib << 1) & 0x4;
+        revnib += (nib << 3) & 0x8;
+
+        return(revnib);
+}
+
+//assumes little endian - find if bit is set to 1 using position 0-7 left to right
+int isNthBitSet (uint8_t c, int n) {
+        static uint8_t  mask[] = {128, 64, 32, 16, 8, 4, 2, 1};
+        return ((c & mask[n]) != 0);
+}
+
+//assumes little endian - usage printBits(sizeof(i), &i);
+void printBits(size_t const size, void const * const ptr)
+{
+        unsigned char *b = (unsigned char*) ptr;
+        unsigned char byte;
+        int i, j;
+
+        for (i=size-1;i>=0;i--)
+        {
+                for (j=7;j>=0;j--)
+                {
+                    byte = (b[i] >> j) & 1;
+                    fprintf(stderr, "%u", byte);
+                }
+        }
+        fprintf(stderr, " ");
+        //puts("");
+}
+
+char * int2bin(uint8_t i)
+{
+        size_t bits = sizeof(uint8_t) * 8;
+
+        char * str = malloc(bits + 1);
```

Comment:
ok, this is rather bad. malloc and then just leaking the memory (in the sprintf's below).
Just remove the whole int2bin business and sprintf("%x... hex.

Reply:
that would print hex.. how do I print binary?

Reply:
there is no easy way to print binary;) you'll get used to see hex as binary in your mind after a while, i guess. The decoder is meant to decode the known pieces of the signal. For development I'd just dump the whole thing (as hex) and fiddle with it in python or something.